jsk_recognition_utils/src/geo/line.cpp

```diff
@@ -198,8 +198,60 @@ namespace jsk_recognition_utils
              origin_[0], origin_[1], origin_[2]);
   }
 
-  void Line::point(double alpha, Eigen::Vector3f& output)
+  void Line::point(double alpha, Eigen::Vector3f& output) const
   {
     output = alpha * direction_ + origin_;
   }
+
+  void Line::commonPerpendicularPoints(const Line& other, Eigen::Vector3f& output1, Eigen::Vector3f& output2) const
+  {
+    if(isParallel(other))
+      return;
+    Eigen::Vector3f vec12 = (other.origin_ - origin_);
+    double d12 = direction_.dot(other.direction_);
+    double alpha1 = (vec12.dot(direction_) - (vec12.dot(other.direction_) * d12)) / (1.0 - d12 * d12);
+    double alpha2 = (- vec12.dot(other.direction_) + (vec12.dot(direction_) * d12)) / (1.0 - d12 * d12);
+    point(alpha1, output1);
+    other.point(alpha2, output2);
+  }
+
+  Line::Ptr Line::commonPerpendicular(const Line& other) const
+  {
+    if(isParallel(other)){
+      Line::Ptr null_ret;
+      return null_ret;
+    }
+    Eigen::Vector3f dir = direction_.cross(other.direction_);
+    Eigen::Vector3f vec12 = (other.origin_ - origin_);
+    double d12 = direction_.dot(other.direction_);
+    double alpha = (vec12.dot(direction_) - (vec12.dot(other.direction_) * d12)) / (1.0 - d12 * d12);
+    Eigen::Vector3f orig;
+    point(alpha, orig);
+    Line::Ptr ret (new Line(dir, orig));
+    return ret;
+  }
+
+  Eigen::Vector3f Line::intersection(const Line& other) const
```

Comment:
Reuse one function in another function like https://github.com/jsk-ros-pkg/jsk_recognition/blob/master/jsk_recognition_utils/src/geo/line.cpp#L75-L86 .
```
Eigen::Vector3f intersection_point = Eigen::Vector3f::Zero();
intersection(other, intersection_point);
return intersection_point;
```